src/lti_provider/default/backend.py

```diff
@@ -0,0 +1,206 @@
+"""This module contains an authentication backend based on LTI launch request."""
+
+import logging
+from typing import Any, Optional
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import PermissionDenied
+from django.http import HttpRequest
+
+from ..lti import LTI
+
+logger = logging.getLogger("lti_provider")
+
+USER_MODEL = get_user_model()
+
+
+class LTIBackend:
+    """
+    Authentication backend used by the lti_provider.default.handlers.success handler.
+    It authenticates a user from a verified LTI request and creates a User if necessary.
+
+    You are encouraged to make your own authentication backend, by overriding this one to
+    add your own domain logic.
+    """
+
+    def authenticate(self, request: Optional[HttpRequest], **kwargs):
+        """
+        Authenticate a user from a LTI request
+
+        Args:
+            request: django http request
+            kwargs: additional parameters
+
+        Returns:
+            An authenticated user or None
+        """
+
+        lti_request = kwargs.get("lti_request")
+        if not lti_request:
+            return None
+
+        if not self.check_lti_request_content(lti_request):
+            return None
+
+        username = self.get_username_from_lti_request(lti_request)
+        logger.debug("User %s authenticated from LTI request", username)
+
+        try:
+            user = USER_MODEL.objects.get(username=username)
+        except USER_MODEL.DoesNotExist:
+            user = self.prepare_new_user(lti_request)
+            user.save()
+            logger.debug("User %s created in database", username)
+        if not user:
```

Comment:
If I'm not mistaken, this case can never happen.

I'm not very confortable with the way you get the username and email several times in different methods.
Why not get them once in the current method as you do above (and raise if you don't find them so the `check_lti_request_content` is not required anymore...) and pass them to the `prepare_new_user` method instead of passing the lti_request?

Reply:
> If I'm not mistaken, this case can never happen.

You're right, i'll remove this test.

> I'm not very confortable with the way you get the username and email several times in different methods.
> Why not get them once in the current method as you do above (and raise if you don't find them so the `check_lti_request_content` is not required anymore...) and pass them to the `prepare_new_user` method instead of passing the lti_request?

The goal of this backend is to provide a base implementation that can be extended to add custom domain logic, and that remains compatible with custom user models.
To do that, I had to separate 3 distinct operations : 

- LTI request validation (from a domain point of view) -> `check_lti_request_content`
- User model retrieval -> `get_username_from_lti_request`
- User model creation  -> `prepare_new_user`

For this implementation, your suggestion is totally legit and avoids duplicating username and email retrieval in the LTI request (even if it's not a performance issue, IMO, it's still duplicate code). But it may not fit to every usage ?
 
Here is an example of usage : 

 - I want to only accept request having "role=Instructor" and a non empty `lis_person_sourcedid` field
 - I have a custom user model that use for username a generated value based on the `lis_person_sourcedid` + an information related to the tool consumer that initiated the request.
 - My User model doesn't have any email field (it's possible, right ?)

I think that we can handle most cases with the 3 distinct method I proposed. They get the LTI request as an argument because it contains everything we need in most cases.

Another approach can be to provide a simpler authentication backend and to rewrite it from scratch to add custom logic. But I thought that the proposed solution was more elegant.

Reply:
FYI the current implementation of this backend in ashley looks like this : https://github.com/openfun/ashley/blob/02e9adee4c85b8c6589dc7e934c280d300d7848c/src/ashley/auth/backend.py

Reply:
Your right, but username is also not required. Any field can be declared as the unique user identifier as defined here: https://github.com/django/django/blob/master/django/contrib/auth/models.py#L356

I think the best approach is probably to make the code as straightforward as possible for the default user as defined in contrib/auth... and let people write their own backend if they have different needs?

Reply:
I end up with lot of duplicate code when writing the ashley's specific backend, that's what I made this refactoring. But if it helps to better understand the example, i'll write a backend as straigthforward as possible.

Reply:
But you can still call the get_username method just that you call it only once? in Django contrib auth they use USERNAME in the generic meaning of "identifier key" so maybe you can as well...

Reply:
Here is the [simplified auth backend](https://github.com/openfun/ashley/blob/b6f4650f5d6963e0ddc95cb8d81fab62a23863c7/src/lti_provider/default/backend.py).